From 37735e7c64ed199d2913305ea1db225a212576de Mon Sep 17 00:00:00 2001
From: Sherik Sensin <sherik.a@telekom-digital.com>
Date: Mon, 6 Jul 2026 15:48:57 +0200
Subject: [PATCH] RDKBNETWOR-98 : Support for untagged VLAN

Reason for change: CPE should be able to support IP connectivity(Static/PPPoE/IPoE) over native interface.

Test Procedure: Performed VlanManager Sanity test.
Risks: None.

Signed-off-by: Sherik Sensin A <sherik.a@telekom-digital.com>
---
 source/TR-181/middle_layer_src/vlan_apis.c | 58 ++++++++++++++++++++--
 1 file changed, 55 insertions(+), 3 deletions(-)

diff --git a/source/TR-181/middle_layer_src/vlan_apis.c b/source/TR-181/middle_layer_src/vlan_apis.c
index 96adcb5..54e0c0a 100644
--- a/source/TR-181/middle_layer_src/vlan_apis.c
+++ b/source/TR-181/middle_layer_src/vlan_apis.c
@@ -360,10 +360,14 @@ void * Vlan_Disable(void *Arg)
     pthread_detach(pthread_self());
 
     pthread_mutex_lock(&vlan_access_mutex);
-    //Set EthLink to False. it will take care UnTagged Created Vlan Interface
-    if (Vlan_SetEthLink(pEntry, FALSE, FALSE) == ANSC_STATUS_FAILURE)
+
+    if (pEntry->VLANId != -1)
     {
-        CcspTraceError(("%s-%d: Failed to Enable EthLink\n", __FUNCTION__, __LINE__));
+        //Set EthLink to False. it will take care UnTagged Created Vlan Interface
+        if (Vlan_SetEthLink(pEntry, FALSE, FALSE) == ANSC_STATUS_FAILURE)
+        {
+            CcspTraceError(("%s-%d: Failed to Disable EthLink\n", __FUNCTION__, __LINE__));
+        }
     }
 
     //Delete Created Tagged Vlan Interface
@@ -402,6 +406,20 @@ void * Vlan_Disable(void *Arg)
         Vlan_DeleteInterface(pEntry);
 #endif
     }
+    else if (pEntry->VLANId == -1)
+    {
+        /* If the VLANID = -1, the VLAN is a bridge, delete the bridge and delete the interface from the bridge */
+        if (strcmp(pEntry->BaseInterface, pEntry->Name) != 0)
+        {
+            if (pEntry->BaseInterface[0] != '\0')
+            {
+                v_secure_system("brctl delif %s %s", pEntry->Name, pEntry->BaseInterface);
+            }
+            v_secure_system("ifconfig %s down", pEntry->Name);
+            v_secure_system("brctl delbr %s", pEntry->Name);
+        }
+    }
+
     pEntry->Status = VLAN_IF_DOWN;
     EthLink_SendVirtualIfaceVlanStatus(pEntry->Path, "Down");
     CcspTraceInfo(("%s - %s:Successfully deleted VLAN interface %s\n", __FUNCTION__, VLAN_MARKER_VLAN_IF_CREATE, pEntry->Name));
@@ -665,6 +683,40 @@ void * Vlan_Enable(void *Arg)
         get_uptime(&uptime);
         pEntry->LastChange  =  uptime;
     }
+    else if (pEntry->VLANId == -1)
+    {
+        /* If the VLANID = -1, the VLAN is a bridge, create the bridge and add the interface to the bridge */
+        if (strcmp(pEntry->BaseInterface, pEntry->Name) != 0)
+        {
+            v_secure_system("ip link show %s > /dev/null 2>&1 || brctl addbr %s", pEntry->Name, pEntry->Name);
+            v_secure_system("brctl addif %s %s 2>/dev/null", pEntry->Name, pEntry->BaseInterface);
+            v_secure_system("ifconfig %s up", pEntry->Name);
+        }
+
+        //Get status of VLAN link
+        status = VLAN_IF_DOWN;
+        while(iIterator < 10)
+        {
+            if (ANSC_STATUS_FAILURE == Vlan_GetTaggedVlanInterfaceStatus(pEntry->Name, &status))
+            {
+                CcspTraceError(("%s-%d: Failed to get Tagged Vlan Interface=%s Status \n", __FUNCTION__, __LINE__, pEntry->Name));
+            }
+
+            if (VLAN_IF_UP == status)
+            {
+                EthLink_SendVirtualIfaceVlanStatus(pEntry->Path, "Up");
+                CcspTraceInfo(("%s-%d: Successfully Updated Vlan Status to WanManager for Interface(%s) \n", __FUNCTION__, __LINE__, pEntry->Name));
+                break;
+            }
+
+            iIterator++;
+            sleep(2);
+            CcspTraceInfo(("%s-%d: Interface Status(%d), retry-count=%d \n", __FUNCTION__, __LINE__, status, iIterator));
+        }
+        long uptime = 0;
+        get_uptime(&uptime);
+        pEntry->LastChange  =  uptime;
+    }
     else
     {
         //Enable EthLink and it will take care Creation of UnTagged Vlan Interface.
-- 
2.43.0

